Adding "Allow/Deny Rules priority for ACL" (#123)
eclipse-zenoh/roadmap: rfcs/ALL/Access Control Rules.md

2. Explicit rules with allow permissions come next, i.e., if request matches anything in the *allow* tree then it will be allowed.
3. The *default_permission* value has least priority. It is applied only if there are no matches in the previous two steps.
    
    Note: if *default_permission* is set to `allow`, then there is no need for setting explicit allow rules, since they will be overlooked by the matcher anyways. The same logic does not apply if *default_permission* is set to `deny` and explicit deny, due to permission priorities (see pseudo-code below for more detail)"

The decision logic is explained by the following pseudocode:

```rust
fn is_allowed(key_expr) -> decision {
   if default_permission == deny {
      if !deny_ketree.match(key_expr) && allow_ketree.match(key_expr) {
            return true;
      } else {
          return false;
      }
   } else { 
      // default_permission = allow
      if deny_ketree.match(key_expr) {
         return false;
      } 
      // no need to check for allow_ketree here
      return true; 
   }
}
```

## Key-Expression Matching

> All requets are macthed on keys and key expressions (for more information on key expressions, check out [Key_Expressions](https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Key%20Expressions.md)) . Therefore, it is important to understand how the key-expression matching works, since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.

Reviewer: ```suggestion
All requests are macthed on keys and key expressions (for more information on key expressions, check out [Key_Expressions](https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Key%20Expressions.md)) . Therefore, it is important to understand how the key-expression matching works, since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.
```

|---------------|---------------|-------|-----|
| test/@demo/a         | test/demo/a         | no   | verbatim missing  | 
| test/demo/a         | test/@demo/a         | no   | verbatim missing | 
| test/@demo/a         | test/@demo/*         | yes   | -  | 
| test/@demo/a         | **            | no   | verbatim missing  | 
| test/@demo/a         | test/*/a         | no   | verbatim missing  | 
| test/@demo/*         | */@demo/a         | no    | partial overlap  | 

If the match happens then the result will be as set in the explicit rules. If not, then the default permission will take over. For example, if the default permission is `allow` and the key-expressions in the *explicit deny* rules are any of the key-expressions like `test/demo/a` , `test/demo/*` or `test/demo/**` then a request on `test/demo/a` will match with the *explicit deny* KeTree and will be denied. However, given the same list of rules, a request on `test/**` will not match (since it is a superset) and therefore the request will be allowed to go through. Therefore, extra care needs to be taken while devising the rules, and especially so when using wildcards.

## Performance

> Given Zenoh's priority is performance, a lot of care was taken while adding access control features to the codebase, to keep the performance as high as possible. However, as with any other piece of software, security comes with a price in terms of performance. However, having done multiple tests, we can share some tips to improve performance:

Reviewer: ```suggestion
Given Zenoh's priority is performance, a lot of care was taken while adding access control features to the codebase, to keep the performance as high as possible. However, as with any other piece of software, security comes with a price in terms of performance. Having done multiple tests, we can share some tips to improve performance:
```

1. Keys (eg: `test/demo/a` ) are faster than key-expressions that use wildcards and DSL (eg: `test/demo/*` or`test/d$*`). Therefore, don't use them in your list of rules unless necessary. Verbatims are okay.
2. The number of chunks in a key-expression also affects the performance since it increases the depth of the KeTree to be searched. So `test/demo/a` will be faster than `test/demo/a/b/c`. This loss of performance is not as drastic as that of using wildcards and DSL. However, still try to keep the number of chunks as low as possible in the key expression.

> 3. Using both flows in the list of rules can results in checking of the messages twice. If possible, use only a single flow in your rules.

Reviewer: ```suggestion
3. Using both flows in the list of rules can cause a double verification of messages. If possible, only use a single flow in your rules.
```

# Access Control Rules 

> Starting from 0.11.0 release, zenoh provides the option of controlling access to key-expressions (eg: `test/demo` ) based on network interface (eg: `lo`). The access control is managed by filtering messages (denoted as **actions** in the acl config): `put`, `get`, `declare_subscriber`, `declare_queryable`. This filteration can be applied on both sides of the interceptor: the incoming messages (`ingress`) and outgoing messages (`egress`).

Reviewer: The concept of an interceptor was not introduced. We should either briefly explain what it does, or link to a page where it is explained.

## Access Control Config

The typical ACL configuration in the config.json5 file looks something like this:

```json5
 acl: {
  ///[true/false] acl will be activated only if this is set to true
  "enabled": false,
  ///[deny/allow] default permission is deny (even if this is left empty or not specified)
  "default_permission": "deny",

>   ///rule set for permissions allowing or denying access to key-expressions

Reviewer: I would say "allowing or denying actions on key-expressions" instead of "allowing or denying access to key-expressions"

The *rules* section itself has 5 inner fields: *actions*, *flows*, *permission*, *key_exprs*, *interfaces*

* **actions**: supports four types of messsges - `put`, `get`, `declare_subscriber`, `declare_queryable`
* **flows**: supports two values - `egress` and `ingress`
* **permission**: supports - `allow` and `deny`
* **key_exprs**: supports values of any key type or key-expression (set of keys) type, eg: `temp/room_1`, `temp/**` etc
* **interfaces**: supports all possible value for network interfaces, eg: `lo`, `lo0` etc

For each combination of *interface* + *flow* + *action*  (example:`l0`+`egress`+`put`) in the rules vector, we construct *allow* and *deny* KeTrees (key-expression tries). The *allow* KeTree is built using all the key-expressions provided in the config on which that *interface* is allowed to do that *action* on a particular *flow*. On receiving an authorization request, the key-expression in the request is matched against the appropriate KeTrees to confirm authorization.

## Priority 
For each message, the access control checks are made in the following order of priority:

*explicit deny* rule > *explicit allow* rule > *default_permission* rule

> 1. *Explicit deny* rules are given topmost priority. This means that if the key-expression in the  authorization request matches against anything in the *deny* KeTree, it will be denied immediately. If there is no match, then the *explicit allow* rules are checked.

Reviewer: It's better to say "explicit rules with `deny` permissions" to avoid confusing permissions with rules.
Same for the following point below regarding `allow` permissions.

2. *Explicit allow* permissions come next, i.e., if request matches anything in the *allow* tree then it will be allowed.
3. The *default_permission* value has least priority. It is applied only if there are no matches in the previous two steps.
    

>     Note: if *default_permission* is set to `allow`, then there is no need for setting *explicit allow* rules, since they will be discarded anyways.

Reviewer: Does the same apply for *default_permission* `deny` and explicit rules with `deny` permissions? From the pseudo-code below I understand that it does not, but we should mention it here too.
ex: "The same logic does not apply to *default_permission*=`deny` and explicit `deny`, due to permission priorities (see pseudo-code below for more detail)"

    

The decision logic is as follows:

```rust
fn is_allowed(key_expr) -> decision {
   if default_permission == deny {
      if !deny_ketree.match(key_expr) && allow_ketree.match(key_expr) {
            return true;
      } else {
          return false;
      }
   } else { //default_permission = allow
      if deny_ketree.match(key_expr) {
         return false;
      } 
			//no need to check for allow_ketree here
			return true; 
   }
}
```

## Key-Expression Matching

> An important thing to note here is how our key-expression matching works since it ultimately decides the behavior of the access control logic. In matching a key-expression against a KeTree, it will match as a positive only if it is *included in* (*equal to* or *subset of*) the key-expressions specified in the KeTree. A partial match or being a superset will not result in a match.

Reviewer: It's a good idea to reference to the [Key Expressions RFC](https://github.com/eclipse-zenoh/roadmap/blob/main/rfcs/ALL/Key%20Expressions.md) file so readers know where to look to learn what are partial matches and supersets.

The following table demonstrates how the matching will work on a key-expression(KE) in request and in the rules set:

| KE in request | KE in rules set | Match |
|---------------|---------------|-------|
| test/demo/a         | test/demo/a         | yes   |
| test/demo/a         | test/demo/*         | yes   |
| test/\*/\*         | test/**          | yes   |
| test/demo/a         | **            | yes   |
| test/demo/a         | test/*/a         | yes   |

> | test/demo/*         | test/*/a         | no    |

Reviewer: Maybe add "(partial match)" next to the `no`to justify why it does not work.
Same for the following line, can add "(superset)"
Same for the other examples of this RFC.

| KE in request | KE in rules set | Match |
|---------------|---------------|-------|
| test/@demo/a         | test/demo/a         | no   |
| test/demo/a         | test/@demo/a         | no   |
| test/@demo/a         | test/@demo/*         | yes   |
| test/@demo/a         | **            | no   |
| test/@demo/a         | test/*/a         | no   |
| test/@demo/*         | */@demo/a         | no    |

If the match happens then the result will be as set in the explicit rules. If not, then the default permission will take over. For example, if the default permission is `allow` and the key-expressions in the *explicit deny* rules are any of the key-expressions like `test/demo/a` , `test/demo/*` or `test/demo/**` then a request on `test/demo/a` will match with the *explicit deny* KeTree and will be denied. However, given the same rules set, a request on `test/**` will not match (since it is a superset) and therefore the request will be allowed to go through. Therefore, extra care needs to be taken while devising the rules, and especially so when using wildcards.

## Performance

> Given, zenoh's priority is performance, a lot of care was taken while adding access control features to the codebase, to keep the performance as high as possible. However, as with any other piece of software, security comes with a price in terms of performance. However, having done multiple tests, we can share some tips to improve performance:

Reviewer: Nitpicking:
- Remove comma after "Given"? Idk it doesn't sound correct to me 😅
- Capitalize "Zenoh"

> 1. Keys (eg: `test/demo/a` ) are faster than key-expressions that use wildcards and DSL (eg: `test/demo/*` or`test/d$*`). Therefore, don't use them in your rules set unless necessary. Verbatims are okay.

Reviewer: Nitpicking again: I feel like "rules set" doesn't sound correct. We should just say "rule set", which means a set of one or multiple rules. If we want to say multiple sets of rules then it should be "rule sets".

Author: changing to "list of rules" as that sounds more appropriate.
